From c69288d759f3f84d2401eefed6542a27e30fd2a4 Mon Sep 17 00:00:00 2001
From: Tim Harvey <tharvey@gateworks.com>
Date: Wed, 11 Mar 2020 08:43:53 -0700
Subject: [PATCH 559/767] gpio: thunderx: fix irq_request_resources

If there are no parent resources do not call irq_chip_request_resources_parent
at all as this will return an error.

This resolves a regression where devices using a thunderx gpio as an interrupt
would fail probing.

Fixes: 0d04d0c ("gpio: thunderx: Use the default parent apis for {request,release}_resources")
Signed-off-by: Tim Harvey <tharvey@gateworks.com>
[Kevin: Original patch taken from:
https://lore.kernel.org/lkml/1583941433-15876-1-git-send-email-tharvey@gateworks.com]
Signed-off-by: Kevin Hao <haokexin@gmail.com>
---
 drivers/gpio/gpio-thunderx.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/drivers/gpio/gpio-thunderx.c b/drivers/gpio/gpio-thunderx.c
index 7463fa1136be..6187ea678828 100644
--- a/drivers/gpio/gpio-thunderx.c
+++ b/drivers/gpio/gpio-thunderx.c
@@ -720,15 +720,18 @@ static int thunderx_gpio_irq_request_resources(struct irq_data *data)
 {
 	struct thunderx_line *txline = irq_data_get_irq_chip_data(data);
 	struct thunderx_gpio *txgpio = txline->txgpio;
+	struct irq_data *parent_data = data->parent_data;
 	int r;
 
 	r = gpiochip_lock_as_irq(&txgpio->chip, txline->line);
 	if (r)
 		return r;
 
-	r = irq_chip_request_resources_parent(data);
-	if (r)
-		gpiochip_unlock_as_irq(&txgpio->chip, txline->line);
+	if (parent_data && parent_data->chip->irq_request_resources) {
+		r = irq_chip_request_resources_parent(data);
+		if (r)
+			gpiochip_unlock_as_irq(&txgpio->chip, txline->line);
+	}
 
 	return r;
 }
-- 
2.31.1

